src/util/archive/fd_tar.c

```diff
@@ -0,0 +1,198 @@
+  reader->header.name[ FD_TAR_NAME_SZ-1 ] = '\0';
+
+  /* Call back to recipient */
+  int err = reader->cb_vt.file( reader->cb_arg, &reader->header, file_sz );
+  return fd_int_if( err, EIO, 0 );
+}
+
+static int
+fd_tar_read_hdr( fd_tar_reader_t * reader,
+                 uchar const **    pcur,
+                 uchar const *     end ) {
+
+  uchar const * cur = *pcur;
+
+  /* Skip padding */
+  ulong  pad_sz = fd_ulong_align_up( reader->pos, 512UL ) - reader->pos;
+         pad_sz = fd_ulong_min( pad_sz, (ulong)( end-cur ) );
+  cur += pad_sz;
+
+  /* Determine number of bytes to read */
+  long chunk_sz = (long)sizeof(fd_tar_meta_t) - (long)reader->buf_ctr;
+  FD_TEST( chunk_sz>=0L );
+  if( end-cur < chunk_sz ) chunk_sz = end-cur;
+
+  /* Copy to header */
+  fd_memcpy( reader->buf + reader->buf_ctr, cur, (ulong)chunk_sz );
+  cur             +=         chunk_sz;
+  reader->buf_ctr += (ushort)chunk_sz;
+
+  /* Handle complete header */
+  int ret = 0;
+  if( reader->buf_ctr == sizeof(fd_tar_meta_t) )
+    ret = fd_tar_process_hdr( reader );
+
+  *pcur = cur;
+  return ret;
+}
+
+static int
+fd_tar_read_data( fd_tar_reader_t * reader,
+                  uchar const **    pcur,
+                  uchar const *     end ) {
+
+  uchar const * cur = *pcur;
+  FD_TEST( cur<=end );
+
+  /* Determine number of bytes to read */
+  ulong chunk_sz = reader->file_sz;
+  ulong avail_sz = (ulong)( end-cur );
+  if( avail_sz < chunk_sz ) chunk_sz = avail_sz;
+
+  /* Call back to recipient */
+  int err = reader->cb_vt.read( reader->cb_arg, cur, chunk_sz );
+
+  /* Consume bytes */
+  cur             += chunk_sz;
+  reader->file_sz -= chunk_sz;
+
+  *pcur = cur;
+  return fd_int_if( err, EIO, 0 );
+}
+
+int
+fd_tar_read( void *        const reader_,
+             uchar const * const data,
+             ulong         const data_sz ) {
+
+  fd_tar_reader_t * reader = reader_;
+  ulong const pos = reader->pos;
+
+  uchar const * cur = data;
+  uchar const * end = cur+data_sz;
+
+  while( cur!=end ) {
+    if( reader->file_sz ) {
+      int err = fd_tar_read_data( reader, &cur, end );
+      if( FD_UNLIKELY( !!err ) ) return err;
+      reader->pos = pos + (ulong)( cur-data );
+    }
+    if( !reader->file_sz ) {
+      int err = fd_tar_read_hdr( reader, &cur, end );
+      if( FD_UNLIKELY( !!err ) ) return err;
+      reader->pos = pos + (ulong)( cur-data );
+    }
+  }
+
+  return 0;
+}
+
+FD_FN_PURE ulong
+fd_tar_meta_get_size( fd_tar_meta_t const * meta ) {
+  char const * buf = meta->size;
+
+  ulong ret = 0UL;
+  for( char const * p=buf; p<buf+12; p++ ) {
+    if( *p == '\0' ) break;
+    ret = (ret << 3) + (ulong)(*p - '0');
+  }
+
+  return ret;
+}
+
+int
+fd_tar_set_octal( char  buf[ static 12 ],
+                  ulong val ) {
+  buf[ 11 ] = '\0';
+  for( int i=10; i>=0; i-- ) {
+    buf[ i ] = '0' + (char)( val&7UL );
+    val>>=3;
+  }
+  return val==0UL;
+}
+
+#include <sched.h>
+
+int
+fd_tar_read_fd( fd_tar_reader_t * reader,
```

Comment:
I kind of hate this read_fd API. There's no file descriptor that you want to spin loop and read() from, even with the sched_yield, you're just going to lock the core up with read() -> EAGAIN a million times.  Has this API been useful? I feel like if the caller is not driving the read loop and you really want this function, it should just issue blocking reads on FD and just park the OS thread.